runner/src/stream-handler/worker.ts

```diff
@@ -19,40 +30,73 @@ void (async function main () {
 
   console.log('Started processing stream: ', streamKey);
 
-  let indexerName = '';
   const streamType = redisClient.getStreamType(streamKey);
-  const isHistorical = streamType === 'historical';
+  isHistorical = (streamType === 'historical');
+
+  await handleStream(streamKey);
+})();
+
+async function handleStream (streamKey: string): Promise<void> {
+  void streamerMessageQueueProducer(queue, streamKey);
+  void streamerMessageQueueConsumer(queue, streamKey);
+}
 
+async function streamerMessageQueueProducer (queue: Array<Promise<QueueMessage>>, streamKey: string): Promise<void> {
   while (true) {
-    try {
-      const startTime = performance.now();
+    const preFetchCount = HISTORICAL_BATCH_SIZE - queue.length;
+    if (preFetchCount <= 0) {
+      await sleep(300); // Wait for more messages in array to process
+      continue;
+    }
+    const messages = await redisClient.getNextStreamMessage(streamKey, preFetchCount);
+    if (messages == null) {
+      await sleep(1000); // Wait for new messages to appear in stream
+      continue;
+    }
+    console.log(`Fetched ${messages?.length} messages from stream ${streamKey}`);
 
-      const messages = await redisClient.getNextStreamMessage(streamKey);
-      const indexerConfig = await redisClient.getStreamStorage(streamKey);
+    for (const streamMessage of messages) {
+      const { id, message } = streamMessage;
+      fetchAndQueue(queue, Number(message.block_height), id);
```

Comment:
```suggestion
      queue.push(transformStreamerMessageToQueueMessage(blockHeight, id));
```
Nit: I think this code is simple/descriptive enough to just be inlined?